src/assets/scripts/client/aircraft/AircraftModel.js

```diff
@@ -1457,18 +1545,48 @@ export default class AircraftModel {
             //     return;
 
             case MCP_MODE.ALTITUDE.VNAV: {
-                const waypointAltitude = this.fms.currentWaypoint.altitudeRestriction;
-                const waypointHasAltitude = waypointAltitude !== -1;
+                const hardRestrictedWaypointModel = this.fms.nextHardAltitudeRestrictedWaypoint;
 
-                if (!waypointHasAltitude) {
-                    return this.mcp.altitude;
+                if (_isNil(hardRestrictedWaypointModel)) {
+                    return;
                 }
 
+                let waypointAltitude = hardRestrictedWaypointModel.altitudeMaximum;
+
                 if (this.flightPhase === FLIGHT_PHASE.TAKEOFF || this.flightPhase === FLIGHT_PHASE.CLIMB) {
```

Comment:
why not use a `switch` for the following logic blocks instead of looking at the same var so many times?  Also, split out the guts of each into separate methods.